### PR TITLE
feat: Initialize Docker

### LINKED ISSUE #1: Starting with the basics
Ok, for better or worse, we have written the Hello World exercise and here we are going to start parallel work for the test runner, but even if I skim through the readme, I get stuck when everything is given all at once and I say, Hey Murat (my real name) how are you going to start now?

So can we start from the basics and give me a basic idea of how to get started?
I'm not very professional especially on how to integrate Docker.

Help wanted from: @ErikSchierboom

### PR DESCRIPTION
**Pull request by GroophyLifefor**
Please don't merge this quickly,

There are still too many mistakes here.
As my `/test-runner/main.bat` executes well 
<kbd>/test-runner/main.bat</kbd>
```cmd
@echo have to say something here.
```
<kbd>Output: </kbd>
```
realpath: _solution: No such file or directory
realpath: _output: No such file or directory
mkdir: can't create directory '': No such file or directory
Greeting: testing...
have to say something here.
Greeting: done
```

<kbd>Results.json</kbd>
```json
{
  "version": 1,
  "status": "fail",
  "message": ""
}
/EOL
```

I'm confused about how to pull tests from the slug.
where are the test-cases kept? where do I get the expected value and the actual value?

### CODE REVIEW

**Comment by GroophyLifefor** [comment]
For example, now the test cases in the workflow have exploded but I'm stuck on how to solve it. I don't know where to pull the slug data from, I hope it doesn't come from the sky.

**Comment by ErikSchierboom** [comment]
@GroophyLifefor I'll have a look for you

**Comment by ErikSchierboom** [comment]
I've looked at the `batch` repo's `hello-world` exercise, but I think we need to fix something there first. If I run `GreeterTest.bat`, it looks like it doesn't actually run the `Greeter.bat` script:

```
C:\Users\erik_\Downloads\hello-world>GreeterTest.bat
Expected: Hello, World!
Actually:
Test failed
```

Could you fix that first? Then I can continue work on this.

**Comment by GroophyLifefor** [comment]
Because test script waits a path parameter but I forget to add an warning as 'usage'.
> C:\Users\erik_\Downloads\hello-world>GreeterTest.bat Greeting.bat

Is the right usage, sorry this my bad.

**Comment by GroophyLifefor** [comment]
Btw I'll fix when I open pc, I'm currently at Gym.

**Comment by ErikSchierboom** [comment]
> Because test script waits a path parameter but I forget to add an warning as 'usage'.

Could you make it so that it doesn't need the path parameter but will just use the correct path itself? I'd like students to just be able to run `HelloWorldTest.bat` (I would also suggest to name the files the same as the exercise slug)

**Comment by GroophyLifefor** [comment]
@ErikSchierboom, I've fixed this, you can check there https://github.com/exercism/batch/pull/6

**Comment by ErikSchierboom** [comment]
@GroophyLifefor Okay, I've looked into this a bit and have made some progress, but I'm not done yet. There were a couple of errors:

1. wine seems to not work on read-only files. We can workaround this for now, but we might require read only on the file system sometime in the future
2. I can't get the tests to run properly.

Obviously the second issue is the biggest problem. When I try to run the test runner, I get this:

```shell
wine cmd /C SuccessTest.bat 
002c:err:winediag:getaddrinfo Failed to resolve your host name IP
Can't recognize 'CMD.EXE /C Z:\solution\Success.bat' as an internal or external command, or batch script
```

This is somewhat weird as when I boot into the docker container interactively (using `docker build --rm -t exercism/batch-test-runner . && docker run -it --entrypoint /bin/sh --rm --network none --mount type=bind,src=$PWD/tests/success,dst=/solution --mount type=bind,src=$PWD/tests/success,dst=/output --mount type=tmpfs,dst=/tmp exercism/batch-test-runner`), I can run `Success.bat` directly, but not indirectly via `SuccessTest.bat`:

```shell
/opt/test-runner # cd /solution
/solution # wine cmd /C Success.bat
Goodbye, Mars!
/solution # 
/solution # wine cmd /C SuccessT.bat
/solution # e
/solution # wine cmd /C SuccessTest.bat
002c:err:winediag:getaddrinfo Failed to resolve your host name IP
Can't recognize 'CMD.EXE /C Z:\solution\Success.bat' as an internal or external command, or batch script.
Expected: Hello, World!
Actually:
Test failed
```

And idea?


/solution # wine cmd /C Success.bat 
002c:err:winediag:getaddrinfo Failed to resolve your host name IP
Goodbye, Mars!

**Comment by GroophyLifefor** [comment]
@ErikSchierboom 
1. I'm not sure how to solve this yet, we can consult with an expert. I also don't know what kind of damage it will do in the long-term. 
2. We can't output the tests properly because I noticed that it can't read the parameter.

run.sh works, but it doesn't have parameters on it, so it keeps printing <kbd>usage: ./bin/run.sh exercise-slug path/to/solution/folder/ path/to/output/directory/</kbd>, if you run it as direct, the output of an empty command is empty because the parameter is empty.

When I try to run it directly, now our tests are in /opt/test-runner and we try to run it with wine from there, and wine runs in /root/wine/drive_c, so even if we have the slug "Success", the part where it runs "SuccessTest.bat" is handled by wine itself, but when we say to find "Success. bat", it looks for it in its own folder in drive_c and cannot find it, so we can access wine from alpine, but wine cannot access alpine, so Success.bat can never be found, there is actually a way to solve this, if we move the test files to the simulated folder where wine is, we will solve the finding issue to a high extent.

I'm waiting for your opinion, do you think it makes sense?

**Comment by ErikSchierboom** [comment]
> there is actually a way to solve this, if we move the test files to the simulated folder where wine is, we will solve the finding issue to a high extent.

Do you know where that is?

**Comment by GroophyLifefor** [comment]
I mean 
```bash
cp /opt/test-runner/tests/Success.bat /root/wine/drive_c/...[a path which we choose]
```

**Comment by ErikSchierboom** [comment]
I've tried that but it still seems to fail:

```
~/.wine/drive_c/users/root/Documents # ls -al
total 32
drwxr-xr-x    1 root     root          4096 Jan 18 11:32 .
drwxr-xr-x    1 root     root          4096 Jan 18 11:33 ..
-rw-r--r--    1 root     root            21 Jan 18 11:32 Success.bat
-rw-r--r--    1 root     root          1234 Jan 18 11:32 SuccessTest.bat
-rw-r--r--    1 root     root            39 Jan 18 11:32 expected_results.json
-rw-r--r--    1 root     root            56 Jan 18 11:32 results.json
drwxr-xr-x    3 root     root          4096 Jan 18 11:32 solution
~/.wine/drive_c/users/root/Documents # whoami
root
~/.wine/drive_c/users/root/Documents # wine cmd /c SuccessTest.bat 
002c:err:winediag:getaddrinfo Failed to resolve your host name IP
Can't recognize 'CMD.EXE /C C:\users\root\Documents\Success.bat' as an internal or external command, or batch sc
ript.
Expected: Hello, World!
Actually:
Test failed
```

**Comment by GroophyLifefor** [comment]
@ErikSchierboom can you commit your current environment, for a continuity without conflict.

**Comment by GroophyLifefor** [comment]
It's interesting, wine don't work well somehow, big probaly we trying something wrong in wine usage case. Maybe we can open an topic in stackoverflow, since we get the correct answer, I'll continue to solve problem, what you think on that, should we open?

**Comment by ErikSchierboom** [comment]
> It's interesting, wine don't work as well somehow, big probaly we trying something wrong in wine usage case. Maybe we can open an topic in stackoverflow, since we get the correct answer, I'll continue to solve problem, what you think on that, should we open?

Yeah I think that's wise (opening an issue). You could also open a topic on the exercism forum too: https://forum.exercism.org/

**Comment by GroophyLifefor** [comment]
I am making the necessary posts.

**Comment by GroophyLifefor** [comment]
https://stackoverflow.com/questions/77846151/apine-wine-cannot-find-the-path-specified-in-the-parameter
https://forum.exercism.org/t/apine-wine-cannot-find-the-path-specified-in-the-parameter/9299

**Comment by iHiD** [comment]
ChatGPT says...

---

It appears that you are encountering an issue where the `SuccessTest.bat` script is unable to correctly execute the `Success.bat` script when run through Wine on an Alpine Linux container. The error message suggests that Wine is unable to recognize or execute the command specified in the `SuccessTest.bat` script.

Here are a few things you can check and try to resolve this issue:

1. **File Paths and Wine Configuration**: Ensure that the file paths used in the script are correctly mapped to Wine's virtual drive structure. In Wine, the `Z:` drive is mapped to the root of the Unix file system. It seems you are copying the files to `~/root/.wine/drive_c/users/root/Documents/`, so ensure that this path is correctly accessible in Wine.

2. **File Permissions**: Make sure that the `.bat` files have the correct permissions set. They need to be executable. You can set this with the command `chmod +x /path/to/your/.bat` files.

3. **Wine Prefix and Environment**: Sometimes, setting up a specific Wine prefix and ensuring that the environment is properly configured can help. You can set up a new Wine prefix and try running your scripts in that environment.

4. **Debugging Wine**: Run Wine with increased verbosity or debugging information to see more detailed error messages. You can do this by setting the `WINEDEBUG` environment variable. For example, `WINEDEBUG=warn+all wine cmd /c SuccessTest.bat` might provide more insight.

5. **Script Syntax**: Double-check the syntax in your batch files. Ensure that the paths and commands are correctly formatted for execution in the Wine environment.

6. **Using Absolute Paths**: In the script, you are using `%~dp0` to refer to the directory of the batch file. While this usually works in a standard Windows environment, there might be quirks with how Wine handles these paths. Try using absolute paths to refer to the `Success.bat` file.

7. **Test with a Simple Batch File**: Create a very simple batch file that just echoes some text and try running it with Wine. This can help determine if the problem is with the Wine setup or with the specific scripts you are using.

8. **Wine Version**: Check if you are using the latest stable version of Wine. Sometimes, updating to the latest version can resolve compatibility issues.

If these suggestions do not resolve the issue, it may be necessary to explore alternative methods to execute your scripts or to debug further within the Wine environment to identify the specific cause of the issue.

**Comment by GroophyLifefor** [comment]
@iHiD Thank you for your attention, unfortunately, I have attempted to fix it using various methods, but have been unsuccessful.

**Comment by GroophyLifefor** [comment]
It's maybe more complicated than we think. To solve those problems may we need a different way which I really don't know.

**Comment by ErikSchierboom** [comment]
@GroophyLifefor I've made some progress. If you use `CALL`, it _does_ work: https://github.com/exercism/batch-test-runner/pull/2/commits/842e187e5df5cfa9df5eba3bc7237a4ce9537d09 I don't know if it is possible to use call and to capture the output in a variable, but this should at least be able to function as a workaround.

**Comment by GroophyLifefor** [comment]
> @GroophyLifefor I've made some progress. If you use `CALL`, it _does_ work: https://github.com/exercism/batch-test-runner/pull/2/commits/842e187e5df5cfa9df5eba3bc7237a4ce9537d09 I don't know if it is possible to use call and to capture the output in a variable, but this should at least be able to function as a workaround.

Ah, It's really powerful trick to fix! Absolutely we can use. I don't used because those are mostly same thing, just one diff was with new method we just create a file, I'll try when I was at home! Thanks you for telling to me.

**Comment by GroophyLifefor** [comment]
@ErikSchierboom This worked very well, there is no problem now. What should we do now? Also, as a request, I would appreciate it if you keep the commands you use to stand up docker in build.bat, it is very important for me and for future contributors.

```
/opt/test-runner # mkdir -p ~/root/.wine/drive_c/users/root/Documents/
/opt/test-runner # cp -R /opt/test-runner/tests/Success/* ~/root/.wine/drive_c/users/root/Documents/
/opt/test-runner # cd ~/root/.wine/drive_c/users/root/Documents/
~/root/.wine/drive_c/users/root/Documents #  wine cmd /c SuccessTest.bat
Expected: Hello, World!
Actually: Goodbye, Mars!
Test failed
~/root/.wine/drive_c/users/root/Documents # ls
Success.bat            SuccessTest.bat        expected_results.json  log.txt
```

File: .dockerignore [hunk 1]
```diff
@@ -7,4 +7,4 @@ Dockerfile
 bin/run-in-docker.sh
 bin/run-tests-in-docker.sh
 bin/run-tests.sh
-tests/
+# tests/
```
Comment (ErikSchierboom) [inline .dockerignore:10]: Could you revert this?

File: tests/success/SuccessTest.bat [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        CALL %slug%.bat > stdout.bin
+        set /p result=<stdout.bin
+        del stdout.bin
+    ) else (
+        REM Errorlevel = 2: The system cannot find the file specified. 
+        REM                 Indicates that the file cannot be found in specified location.
+        echo Could not find "%~dp0%slug%.bat" or "%slug%.bat"
+        exit /b 2
+    )
+)
```
Comment (ErikSchierboom) [inline tests/success/SuccessTest.bat:36]: I don't fully understand why we need to relatively complex behavior. When the student downloads the exercise it is guaranteed that the test and solution files are downloaded, so you can just assume that the solution file is there next to the test file. Having the test file be simpler is also really useful for the student.

**Review by ErikSchierboom — COMMENTED** [review]
> Also, as a request, I would appreciate it if you keep the commands you use to stand up docker in build.bat, it is very important for me and for future contributors. 

What is the goal of using a docker compose file here? You should only really have to run `./bin/run-in-docker.sh` or `./bin/run-tests-in-docker.sh`.

> What should we do now

Could you update the other test cases' test files too?

File: .dockerignore (hunk 1 above)
Comment (GroophyLifefor) [reply .dockerignore:10]: In the next coming commit.

File: tests/success/SuccessTest.bat (hunk 2 above)
Comment (GroophyLifefor) [reply tests/success/SuccessTest.bat:36]: In fact, but a second possibility was to find the way, but it may not be so necessary, now I have written a code that more students can understand, it is completely open to criticism.

**Comment by GroophyLifefor** [comment]
> What is the goal of using a docker compose file here? You should only really have to run ./bin/run-in-docker.sh or ./bin/run-tests-in-docker.sh.

Maybe not all people use unix, after all windows is very common, but that's ok, I'll delete the build.bat for now.

> Could you update the other test cases' test files too?

if you about 'SuccessTest.bat', It's changed in next commit, if you about other files, I don't fully understand which files you talk about.

File: tests/success/SuccessTest.bat (hunk 2 above)
Comment (ErikSchierboom) [reply tests/success/SuccessTest.bat]: I would replace the code with:

```
CALL %slug%.bat > stdout.bin
set /p result=<stdout.bin
del stdout.bin
```

**Comment by ErikSchierboom** [comment]
> if you about 'SuccessTest.bat', It's changed in next commit, if you about other files, I don't fully understand which files you talk about.

There are four other test cases in the `tests` folder, which all have a `*Test.bat` file that will also need to be updated.

File: tests/success/SuccessTest.bat (hunk 2 above)
Comment (iHiD) [reply tests/success/SuccessTest.bat]: @GroophyLifefor Be aware that no student will ever seen any of this. It runs on our servers - it's not something a student would ever download etc :)

**Comment by GroophyLifefor** [comment]
@ErikSchierboom review requested. Also can you look why CI/CD failing?

image

**Comment by ErikSchierboom** [comment]
Sure! I'll look into it

**Comment by ErikSchierboom** [comment]
@GroophyLifefor Could you remove this bit from the test file:

```
REM Export result as JSON
if %status%==pass (
(
  echo {
  echo   "version": 1,
  echo   "status": "%status%"
  echo }
) > results.json
) else (
(
  echo {
  echo   "version": 1,
  echo   "status": "%status%",
  echo   "message": "%message%"
  echo }
) > results.json
)
```

and replace it with a call that exits with the right error code? The `results.json` file is only needed for the test runner, but the tests file will also be run by students locally, who don't need that file.

**Comment by GroophyLifefor** [comment]
@ErikSchierboom I tried to find the document with the correct exit code but I couldn't find it, can you provide a reference link?

**Comment by ErikSchierboom** [comment]
It should be 0 for a successful run (all tests pass), and non-zero for all other cases

**Comment by GroophyLifefor** [comment]
Ah okey, I'll when I was at home, during the day, even though school is on vacation now, I can't make changes quickly because I go to the gym and the library.

**Comment by GroophyLifefor** [comment]
@ErikSchierboom I do, can you please check if everything looks okay?

**Comment by ErikSchierboom** [comment]
I'm not working today so it'll be Tuesday when I have time, sorry

**Comment by ErikSchierboom** [comment]
Okay nice, we're getting there! The `tests/success` case is now working well. Here are the things that we should improve:

1. Don't check for an empty file in the test file

The test files contain logic to detect empty files:

```
:CheckEmptyFile
REM It's for initialize, not about checking empty file
set successCount=0
set failCount=0

for %%I in (%slug%.bat) do (
    if %%~zI equ 0 (
        set "isEmpty=true"
    ) else (
        set "isEmpty=false"
    )
)
GOTO :EOF REM Go back to the line after the call to :CheckEmptyFile
```

I don't think this is something that we need to handle explicitly with the tests file. The reason the test case is there is to verify that an error (where none of the tests could be run) is handled correctly. In this case the error should be something like:

```json
{
  "version": 1,
  "status": "error",
  "message": "???"
}
```

where the message should be something that would help the student figure things out (usually something like "function X could not be found or something like that).


2. The output for failing tests should be more descriptive

Right now, the output looks like:

```
Test failed
Test failed
```

But this is a bit unhelpful for students, as they need to know _which_ tests are failing. 

3. Disable networking
The tests are somewhat slow-ish to execute, and I'm almost positive this is due to wine trying to doing some networking:

```
{
  "version": 1,
  "status": "fail",
  "message": "002c:err:winediag:getaddrinfo Failed to resolve your host name IP\nTest failed\r\nTest failed\r"
}
```

Could you check if it is possible to disable networking at all? https://forum.winehq.org/viewtopic.php?t=30278 suggest that this might be the case.

**Comment by GroophyLifefor** [comment]
I am very proud that we are making progress. 

1. I can delete it, but if I delete this logic, we will no longer know if the file is empty or not, so there will be no special exit code.
2. Fixed in coming commit. I didn't have much time because it was night and I applied it in a single file, if you approve, I will continue.
3. I was interested in the last comment in your reference, it seemed very simple and easy to use, but I haven't tested it because I have some questions on the docker part, I use windows as a host machine so I need to uninstall docker first, but I'm stuck on the part of doing a test with a Slug call, I need to figure it out, can you test the change in run.sh for now?

File: bin/run.sh [hunk 3]
```diff
@@ -37,7 +37,12 @@ cd "${solution_dir}"
 # Run the tests for the provided implementation file and redirect stdout and
 # stderr to capture it
 test_file=$(jq -r '.files.test[0]' "${solution_dir}/.meta/config.json")
-test_output=$(wine cmd /c "${test_file}" 2>&1)
+
+# Run wine with block networking to prevent the tests from making network calls
+test_output=$(sudo unshare -n runuser fabian -c "wine cmd /c \"${test_file}\"" 2>&1)
```
Comment (ErikSchierboom) [inline bin/run.sh:42]: This doesn't work in docker, as there is no `sudo` command. Can this be done in the `Dockerfile` instead somehow?

**Comment by ErikSchierboom** [comment]
> 1. I can delete it, but if I delete this logic, we will no longer know if the file is empty or not, so there will be no special exit code.
Fixed in coming commit. I didn't have much time because it was night and I applied it in a single file, if you approve, I will continue.
I think removing it makes sense. Submitting an empty file is an edge case and only happens rarely. I'd much rather the tests file be smaller and more readable instead.

> 2. Fixed in coming commit. I didn't have much time because it was night and I applied it in a single file, if you approve, I will continue.

Yes, that is lovely! It would be even better if there was something like: "Expected value to be X, but got Y." or something like that. Would that be possible?

> 3. I was interested in the last comment in your reference, it seemed very simple and easy to use, but I haven't tested it because I have some questions on the docker part, I use windows as host machine so I need to setup docker first, but I'm stuck on the part of doing a test with a Slug call, I need to figure it out, can you test the change in run.sh for now?

Unfortunately it doesn't work (see my comment on the PR).

File: bin/run.sh (hunk 3 above)
Comment (GroophyLifefor) [reply bin/run.sh:42]: Ah right, I didn't realize it because I couldn't try it, I found a resource like this, maybe it can really solve our problem, that is, it can speed up the process. https://superuser.com/questions/249962/how-can-i-disable-internet-access-for-programs-running-in-wine
 
By the way, let's not forget that the batch language is already a bit cumbersome.

**Comment by GroophyLifefor** [comment]
1. Okey I'll update when I at home maybe evening.

2. Awesome, I'll look what can I do.

3. I write something as reply in your review.

File: bin/run.sh (hunk 3 above)
Comment (ErikSchierboom) [reply bin/run.sh:42]: I haven't had any luck with that yet

Comment (GroophyLifefor) [reply bin/run.sh:42]: We close the network when run docker, shouldn't wine's requests be rejected anyway?

Comment (ErikSchierboom) [reply bin/run.sh:42]: Yes but it seems to hang. There is something that I can try

Comment (GroophyLifefor) [reply bin/run.sh:42]: > Yes but it seems to hang. There is something that I can try

How did your trials turn out?

Comment (ErikSchierboom) [reply bin/run.sh]: I managed to fix it using something we use for a couple of other test runners too: using an internal network

**Comment by GroophyLifefor** [comment]
It's great that everything is okay now. After that, what are the steps we need to take to upload the Hello world exercise to exercism?

**Comment by ErikSchierboom** [comment]
I think there are a couple of things we need to do to get the test runner in proper shape:

- Simplify the tests file as much as we can. That includes removing the special case handling of empty files.
- Would it be possible to detect when the tests couldn't even execute? This should be the case for the `empty` and `syntax-error` test cases. If none of the tests could be executed, the resulting status should be `error` (and not `fail`).

I would also like to have the test failure messages to include something like "Expected value to be X, but got Y.", but that is something that can be done later.

**Comment by GroophyLifefor** [comment]
- It's almost done, I'll commit it soon. I will also see how else I can simplify it.
- Batch is a slightly different language, in scripting we can detect the error of a code. To be more understandable;

```cmd
REM command
xcurl -p -t
if %errorlevel% neq 0 (
    REM got error 

    REM if errorlevel was 9009, It means not recognized as an internal or external command error.
    REM STDERR: 'xcurl' is not recognized as an internal or external command, operable program or batch file.
)
```

Even though this is possible, after executing this file, when we look at the error level, we will see that it says 0 because even if there is an error inside, when we look at the general, yes, we have executed this file, even if it is faulty, so there is no problem. So the errorlevel will be 0 for us.

There are different ways, what in other languages means 'or' and 'and' is slightly different in batchfile.

```cmd
REM and yes, in batch we can store everything in strings too and executing as well.
set "command=xcurl -p -t" 
REM && means if before command successfully executed.
%command% && echo success.

REM || means if before command failed in executing.
%command% || echo failed.
```

In this case, we can use like this one,

```cmd
CALL %slug%.bat %~1 %~2 %~3 %~4 %~5 %~6 %~7 %~8 %~9 > stdout.bin 2>&1
set /p result=<stdout.bin
del stdout.bin
```

to

```cmd
set "_failed=false"
CALL %slug%.bat %~1 %~2 %~3 %~4 %~5 %~6 %~7 %~8 %~9 > stdout.bin 2>&1 || set "_failed=true"
set /p result=<stdout.bin
del stdout.bin
```

But in this case it can be a bit difficult to understand the problem since we don't have the error code.

**Comment by ErikSchierboom** [comment]
> But in this case it can be a bit difficult to understand the problem since we don't have the error code.

Yeah that is hard. Maybe we can just use some generic error message in that case

**Comment by GroophyLifefor** [comment]
We have an important piece of information, if || is active, something is really wrong with the code.

**Comment by GroophyLifefor** [comment]
Sorry for delay, now a days I'm always out of home because of so much reason. I'll taking care of as soon as.

File: tests/all-fail/AllFailTest.bat [hunk 4]
```diff
@@ -7,21 +7,26 @@ REM ---------------------------------------------------
...
     set "expected=false"
     set "if_success=Test passed"
-    set "if_failed=Test failed: 2015 shouldn't be divided into four."
+    set "if_failed=Test failed: '2015' shouldn't be divided into four."
     CALL :Assert 2015
 
+    set "expected=false"
+    set "if_success=Test passed"
+    set "if_failed=Test failed: Expected value to be '^%expected%', but got '^%stdout%'."
+    CALL :Assert 2017
```
Comment (GroophyLifefor) [inline tests/all-fail/AllFailTest.bat:28]: https://github.com/exercism/batch-test-runner/pull/2#issuecomment-1918635455

> It would be even better if there was something like: "Expected value to be X, but got Y." or something like that. Would that be possible?

**Comment by GroophyLifefor** [comment]
@ErikSchierboom can you review?

**Comment by GroophyLifefor** [comment]
@ErikSchierboom 
    Hi Erik, are you okay? I heard you was sick but I was worried about you.

**Comment by ErikSchierboom** [comment]
I'm almost better, now just dealing with a heavy cold. I'll do a review!

**Review by ErikSchierboom — APPROVED** [review]
I think this is good enough for a first version! Let's merge this.

**Comment by ErikSchierboom** [comment]
The next step would be to apply what we've done here (as in: how to structure the exercise with the tests and such) to the batch repo's exercises.

**Comment by GroophyLifefor** [comment]
> I'm almost better, now just dealing with a heavy cold. I'll do a review!

I'm happy to hear that, take care of yourself. 🧡
